File: firebase-functions/src/handshake_logger.ts
/**
 * Handshake Logger
 *
 * Logs the Firebase/Node.js → Python handshake flow in a readable format.
 * Shows each step of the communication process.
 *
 * Usage:
 *   const logger = new HandshakeLogger("validate-dataset");
 *   logger.logRequestSent(request);
 *   logger.logResponseReceived(response);
 *   logger.logSummary(true, duration);
 */

import {ValidateDatasetRequest, EvaluateModelRequest} from "./ml_runtime_client";

// ============================================================================
// Handshake Logger
// ============================================================================

export class HandshakeLogger {
  private operationName: string;
  private startTime: number = Date.now();
  private logs: string[] = [];
  private useColors: boolean = true;

  // ANSI color codes
  private colors = {
    reset: "\x1b[0m",
    bright: "\x1b[1m",
    dim: "\x1b[2m",
    green: "\x1b[32m",
    red: "\x1b[31m",
    yellow: "\x1b[33m",
    cyan: "\x1b[36m",
  };

  constructor(operationName: string, useColors: boolean = true) {
    this.operationName = operationName;
    this.useColors = useColors;
  }

  /**
   * Log request being sent from Node.js.
   */
  logRequestSent(request: ValidateDatasetRequest | EvaluateModelRequest): void {
    const timestamp = new Date().toISOString();
    const msg = `[${timestamp}] [NODE  ] → Sending request to Flask...`;
    this._log(msg);

    // Extract key fields for display
    const fields: string[] = [];
    if ("dataset_path" in request) {
      const path = request.dataset_path.split("/").pop() || request.dataset_path;
      fields.push(`dataset=${path}`);
    }
    if ("model_path" in request) {
      const path = request.model_path.split("/").pop() || request.model_path;
      fields.push(`model=${path}`);
    }
    if (request.run_id) {
      fields.push(`run_id=${request.run_id}`);
    }

    const fieldStr = fields.join(", ");
    const detailMsg = `${" ".repeat(35)} (${fieldStr})`;
    this._log(detailMsg, "dim");
  }

  /**
   * Log request validated on Flask side.
   */
  logRequestValidated(isValid: boolean): void {
    const timestamp = new Date().toISOString();
    const status = isValid ? "✓" : "✗";
    const statusColor = isValid ? "green" : "red";
    const result = isValid ? "Request valid" : "Request invalid";

    const msg = `[${timestamp}] [FLASK ] ${this._colorize(status, statusColor)} ${result}`;
    this._log(msg);
  }

  /**
   * Log mock response being generated.
   */
  logResponseGenerated(endpoint: string): void {
    const timestamp = new Date().toISOString();
    const endpointName = endpoint.split("/").pop() || endpoint;
    const msg = `[${timestamp}] [FLASK ] ⚙ Generating mock response for ${endpointName}...`;
    this._log(msg);
  }

  /**
   * Log response being returned from Flask.
   */
  logResponseReturned(statusCode: number): void {
    const timestamp = new Date().toISOString();
    const statusColor = statusCode === 200 ? "green" : "red";
    const msg = `[${timestamp}] [FLASK ] ← Returning ${this._colorize(statusCode.toString(), statusColor)} response`;
    this._log(msg);
  }

  /**
   * Log response received on Node.js side.
   */
  logResponseReceived(statusCode: number): void {
    const timestamp = new Date().toISOString();
    const statusColor = statusCode === 200 ? "green" : "red";
    const msg = `[${timestamp}] [NODE  ] ← Received ${this._colorize(statusCode.toString(), statusColor)} response from Flask`;
    this._log(msg);
  }

  /**
   * Log response validated on Node.js side.
   */
  logResponseValidated(isValid: boolean): void {
    const timestamp = new Date().toISOString();
    const status = isValid ? "✓" : "✗";
    const statusColor = isValid ? "green" : "red";
    const result = isValid ? "Response valid" : "Response invalid";

    const msg = `[${timestamp}] [NODE  ] ${this._colorize(status, statusColor)} ${result}`;
    this._log(msg);
  }

  /**
   * Log error that occurred.
   */
  logError(error: string): void {
    const timestamp = new Date().toISOString();
    const msg = `[${timestamp}] [ERROR ] ${this._colorize("✗", "red")} ${error}`;
    this._log(msg);
  }

  /**
   * Log summary of handshake.
   */
  logSummary(success: boolean, durationMs: number): void {
    const timestamp = new Date().toISOString();
    const status = success
      ? this._colorize("✅ SUCCESS", "green")
      : this._colorize("❌ FAILED", "red");
    const msg = `[${timestamp}] [SUMMARY] ${status} (${durationMs}ms)`;
    this._log(msg);
  }

  /**
   * Print all logs.
   */
  printLogs(): void {
    console.log("");
    this.logs.forEach((log) => console.log(log));
    console.log("");
  }

  /**
   * Get all logs as array.
   */
  getLogs(): string[] {
    return this.logs;
  }

  /**
   * Private helper to log with optional color.
   */
  private _log(message: string, colorKey?: keyof typeof this.colors): void {
    if (colorKey && this.useColors) {
      const colorCode = this.colors[colorKey];
      const coloredMsg = `${colorCode}${message}${this.colors.reset}`;
      this.logs.push(message); // Store uncolored version
      console.log(coloredMsg);
    } else {
      this.logs.push(message);
      console.log(message);
    }
  }

  /**
   * Private helper to colorize text.
   */
  private _colorize(text: string, colorKey: keyof typeof this.colors): string {
    if (!this.useColors) {
      return text;
    }
    const colorCode = this.colors[colorKey];
    return `${colorCode}${text}${this.colors.reset}`;
  }
}

// ============================================================================
// Standalone Logging Functions (for simple use cases)
// ============================================================================

/**
 * Log a complete handshake in compact format.
 *
 * @example
 *   logHandshakeFlow({
 *     operation: "validate-dataset",
 *     requestSent: true,
 *     requestValid: true,
 *     responseReceived: true,
 *     responseValid: true,
 *     duration: 245,
 *     success: true
 *   });
 */
export function logHandshakeFlow(options: {
  operation: string;
  requestSent: boolean;
  requestValid: boolean;
  responseReceived: boolean;
  responseValid: boolean;
  duration: number;
  success: boolean;
  error?: string;
}): void {
  const timestamp = new Date().toISOString();
  const status = options.success ? "✅" : "❌";
  const shortOp = options.operation.split("-").pop() || options.operation;

  // Compact single-line format
  const requestStatus = options.requestValid ? "REQ✓" : "REQ✗";
  const responseStatus = options.responseValid ? "RESP✓" : "RESP✗";
  const line =
    `${status} [${shortOp}] ${requestStatus} ${responseStatus} → ${options.duration}ms`;

  console.log(line);

  if (options.error) {
    console.log(`   └─ Error: ${options.error}`);
  }
}

// ============================================================================
// Flow Logger Helper
// ============================================================================

/**
 * Create a standard handshake log with typical flow.
 */
export function createHandshakeLog(): {
  logger: HandshakeLogger;
  startTime: number;
  logFlow: (
    success: boolean,
    requestValid: boolean,
    responseValid: boolean
  ) => void;
} {
  const logger = new HandshakeLogger("handshake");
  const startTime = Date.now();

  const logFlow = (
    success: boolean,
    requestValid: boolean,
    responseValid: boolean
  ) => {
    if (!success) {
      logger.logError("Handshake failed");
    }
    const duration = Date.now() - startTime;
    logger.logSummary(success && requestValid && responseValid, duration);
  };

  return {logger, startTime, logFlow};
}

export default HandshakeLogger;

File: ml-pipeline/src/handshake_logger.py
"""
Handshake Logger

Logs the Firebase/Node.js → Python handshake flow in a readable format.
Shows each step of the communication process on the Flask side.

Usage:
    logger = HandshakeLogger("validate-dataset")
    logger.log_request_received(request)
    logger.log_response_returned(200)
    logger.log_summary(True, duration)
"""

from datetime import datetime
from typing import Any, Dict, List, Optional
import json


class HandshakeLogger:
    """Logger for Flask/Python handshake communication."""

    # ANSI color codes
    GREEN = "\033[92m"
    RED = "\033[91m"
    YELLOW = "\033[93m"
    CYAN = "\033[36m"
    RESET = "\033[0m"
    DIM = "\033[2m"
    BRIGHT = "\033[1m"

    def __init__(self, operation_name: str, use_colors: bool = True):
        """Initialize logger."""
        self.operation_name = operation_name
        self.use_colors = use_colors
        self.logs: List[str] = []
        self.start_time = datetime.now()

    def _colorize(self, text: str, color: str) -> str:
        """Add color to text if enabled."""
        if not self.use_colors:
            return text
        return f"{color}{text}{self.RESET}"

    def log_request_received(self, endpoint: str) -> None:
        """Log request received by Flask."""
        timestamp = datetime.now().isoformat()
        msg = f"[{timestamp}] [FLASK ] ← Request received at {endpoint}"
        self._log(msg)

    def log_request_validating(self) -> None:
        """Log request validation starting."""
        timestamp = datetime.now().isoformat()
        msg = f"[{timestamp}] [FLASK ] ⚙ Validating request..."
        self._log(msg)

    def log_request_validated(self, is_valid: bool, errors: Optional[List[str]] = None) -> None:
        """Log request validation result."""
        timestamp = datetime.now().isoformat()

        if is_valid:
            status = "✓"
            status_color = self.GREEN
            result = "Request valid"
        else:
            status = "✗"
            status_color = self.RED
            result = "Request invalid"

            if errors:
                first_error = errors[0] if isinstance(errors, list) else str(errors)
                result += f" ({first_error})"

        colored_status = self._colorize(status, status_color)
        msg = f"[{timestamp}] [FLASK ] {colored_status} {result}"
        self._log(msg)

    def log_processing(self, processor: str) -> None:
        """Log mock processing happening."""
        timestamp = datetime.now().isoformat()
        msg = f"[{timestamp}] [FLASK ] ⚙ Processing with {processor}..."
        self._log(msg)

    def log_response_generated(self, status_code: int, status_text: str = "") -> None:
        """Log response being generated."""
        timestamp = datetime.now().isoformat()
        status_display = f"{status_code}" + (f" {status_text}" if status_text else "")
        msg = f"[{timestamp}] [FLASK ] ✓ Response generated ({status_display})"
        self._log(msg)

    def log_response_returned(self, status_code: int) -> None:
        """Log response being returned to caller."""
        timestamp = datetime.now().isoformat()

        if status_code == 200:
            status_color = self.GREEN
            status_display = self._colorize(str(status_code), status_color)
        else:
            status_color = self.RED
            status_display = self._colorize(str(status_code), status_color)

        msg = f"[{timestamp}] [FLASK ] → Returning {status_display} response"
        self._log(msg)

    def log_error(self, error_msg: str) -> None:
        """Log error that occurred."""
        timestamp = datetime.now().isoformat()
        colored_status = self._colorize("✗", self.RED)
        msg = f"[{timestamp}] [FLASK ] {colored_status} Error: {error_msg}"
        self._log(msg)

    def log_summary(self, success: bool, duration_ms: float) -> None:
        """Log summary of handshake."""
        timestamp = datetime.now().isoformat()

        if success:
            status = self._colorize("✅ SUCCESS", self.GREEN)
        else:
            status = self._colorize("❌ FAILED", self.RED)

        msg = f"[{timestamp}] [SUMMARY] {status} ({duration_ms:.0f}ms)"
        self._log(msg)

    def _log(self, message: str) -> None:
        """Log message and store it."""
        self.logs.append(message)
        print(message)

    def get_logs(self) -> List[str]:
        """Get all logged messages."""
        return self.logs

    def print_summary(self) -> None:
        """Print all logs together."""
        print("\n".join(self.logs))


def log_handshake_flow(
    operation: str,
    request_valid: bool,
    response_status: int,
    duration_ms: float,
    success: bool,
    error: Optional[str] = None,
) -> None:
    """
    Log a complete handshake in compact format.

    Example:
        log_handshake_flow(
            operation="validate-dataset",
            request_valid=True,
            response_status=200,
            duration_ms=245,
            success=True
        )
    """
    timestamp = datetime.now().isoformat()
    status = "✅" if success else "❌"
    short_op = operation.split("-")[-1] if "-" in operation else operation

    request_status = "REQ✓" if request_valid else "REQ✗"
    response_status_str = "RESP✓" if response_status == 200 else "RESP✗"

    line = (
        f"{status} [{short_op}] {request_status} {response_status_str} → {duration_ms:.0f}ms"
    )
    print(line)

    if error:
        print(f"   └─ Error: {error}")


# ============================================================================
# Example Usage (for testing)
# ============================================================================

if __name__ == "__main__":
    print("=" * 70)
    print("EXAMPLE 1: Successful Handshake")
    print("=" * 70)

    logger1 = HandshakeLogger("validate-dataset")

    logger1.log_request_received("/api/ml/validate-dataset")
    logger1.log_request_validating()
    logger1.log_request_validated(True)
    logger1.log_processing("MockPythonRuntime")
    logger1.log_response_generated(200, "OK")
    logger1.log_response_returned(200)
    logger1.log_summary(True, 245.5)

    print("\n" + "=" * 70)
    print("EXAMPLE 2: Failed Handshake (Invalid Request)")
    print("=" * 70)

    logger2 = HandshakeLogger("evaluate")

    logger2.log_request_received("/api/ml/evaluate")
    logger2.log_request_validating()
    logger2.log_request_validated(False, ["model_path is required"])
    logger2.log_response_returned(400)
    logger2.log_summary(False, 45.2)

    print("\n" + "=" * 70)
    print("EXAMPLE 3: Compact Single-Line Format")
    print("=" * 70)

    log_handshake_flow(
        operation="validate-dataset",
        request_valid=True,
        response_status=200,
        duration_ms=245,
        success=True,
    )

    log_handshake_flow(
        operation="evaluate",
        request_valid=False,
        response_status=400,
        duration_ms=45,
        success=False,
        error="Missing required field: model_path",
    )
